File: day01/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>

auto find(const std::string& line, bool check_words=false) {

    static const std::vector<char> chars = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'};
    static const std::vector<std::string> words = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

    int64_t first = line.size();
    int64_t first_value = 0;
    int64_t last = -1;
    int64_t last_value = 0;

    for(int idx=0; idx < chars.size(); ++idx) {
        int64_t pos = line.find(chars[idx]);
        if (pos!=std::string::npos && pos < first) {
            first = pos;
            first_value = idx;
        }

        pos = line.rfind(chars[idx]);
        if (pos!=std::string::npos && last < pos)
        {
            last = pos;
            last_value = idx;
        }

        if(check_words) {
            int64_t pos = line.find(words[idx]);
            if (pos!=std::string::npos && pos < first) {
                first = pos;
                first_value = idx;
            }

            pos = line.rfind(words[idx]);
            if (pos!=std::string::npos && last < pos)
            {
                last = pos;
                last_value = idx;
            }
        }
    }

    //std::cout << line << " -- " << first_value << ", " << last_value << std::endl;
    return std::make_pair<>(first_value, last_value);
}


int main() {

    {
        std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
        std::string line;

        int64_t total = 0;
        while(std::getline(ifs, line))
        {
            auto first_last = find(line);
            total += (first_last.first*10 + first_last.second);
        }

        std::cout << "Part1: " << total << std::endl;
    }

    {
        std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
        std::string line;

        int64_t total = 0;
        while(std::getline(ifs, line))
        {
            auto first_last = find(line, true);
            total += (first_last.first*10 + first_last.second);
        }


        std::cout << "Part2: " << total << std::endl;
    }

    return 0;
}

File: day01/src/main.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::iter::Map;
use std::path::Path;
use std::slice::Iter;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>> where P: AsRef<Path> {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "input.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn find(line: &str, check_words: bool) -> (usize, usize) {
    let chars = vec!['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
    let words = vec!["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];

    let chars_first = chars.iter().enumerate().filter(|elem| line.find(*elem.1).is_some()).map(|elem| (line.find(*elem.1).unwrap(), elem.0)).min();
    let chars_last = chars.iter().enumerate().filter(|elem| line.find(*elem.1).is_some()).map(|elem| (line.rfind(*elem.1).unwrap(), elem.0)).max();

    if !check_words {
        return (chars_first.unwrap().1, chars_last.unwrap().1);
    }

    let words_first = words.iter().enumerate().filter(|elem| line.find(*elem.1).is_some()).map(|elem| (line.find(*elem.1).unwrap(), elem.0)).min();
    let words_last = words.iter().enumerate().filter(|elem| line.find(*elem.1).is_some()).map(|elem| (line.rfind(*elem.1).unwrap(), elem.0)).max();

    if words_first.is_none() {
        return (chars_first.unwrap().1, chars_last.unwrap().1);
    }


    return (chars_first.min(words_first).unwrap().1, chars_last.max(words_last).unwrap().1);
}

fn main() {
    let mut total: usize = 0;
    for line in read() {
        let (first, last) = find(&line, false);
        // println!("{} -- {}, {}", line, first, last);
        total += first * 10 + last;
    }

    println!("Part1: {}", total);

    let mut total: usize = 0;
    for line in read() {
        let (first, last) = find(&line, true);
        // println!("{} -- {}, {}", line, first, last);
        total += first * 10 + last;
    }

    println!("Part2: {}", total);
}

